### PR TITLE
[Breaking][jvm-packages] add barrier execution mode

### PR DESCRIPTION
**Pull request by wbo4958**
To fix https://github.com/dmlc/xgboost/issues/7835. Credit to @[WeichenXu123](https://github.com/WeichenXu123)

By introducing the barrier execution mode, we don't need to kill SparkContext when some xgboost tasks failed, instead, Spark will abort the whole barrier stage which does not depend on SparkListener, we will never encounter the xgboost task hang issue. So in this PR, the killSparkContextOnWorkerFailure parameter is deleted.

Since this PR has deleted Spark SparkParallelismTracker, the timeoutRequestWorkers parameter is not needed anymore. But one test `cross-version model loading (0.82)` will fail if we just delete timeoutRequestWorkers, we just keep this parameter in this PR. Will file a following up PR to delete timeoutRequestWorkers parameters.

### CODE REVIEW

**Comment by wbo4958** [comment]
@trivialfis could you help to review it?

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala [hunk 1]
```diff
@@ -375,13 +374,14 @@ class XGBoostGeneralSuite extends FunSuite with TmpFolderPerSuite with PerTest {
...
-    val trainingDF = buildDataFrame(Seq(XGBLabeledPoint(1.0f, 1, Array(), Array())))
+    val trainingDF = buildDataFrame(Seq(XGBLabeledPoint(1.0f, 4,
+      Array(0, 1, 2, 3), Array(0, 1, 2, 3))))
     val xgb = new XGBoostClassifier(paramMap)
-    intercept[XGBoostError] {
-      val model = xgb.fit(trainingDF)
+    intercept[SparkException] {
+      xgb.fit(trainingDF)
     }
   }
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala:386]: Are these changes strictly related to the support of barrier mode?

**Review by trivialfis — COMMENTED** [review]
Thank you for working on this, being able to remove code and improve robustness is exciting!

- Could you please share more details on what's the breaking change and how users should adapt to it?
- Could you please keep some RABIT mock tests here to show that exceptions are indeed being well handled?

**Comment by wbo4958** [comment]
> Thank you for working on this, being able to remove code and improve robustness is exciting!
> 
> * Could you please share more details on what's the breaking change and how users should adapt to it?

killSparkContextOnWorkerFailure parameter is deleted. Users are supposed not to use this parameter.

> * Could you please keep some RABIT mock tests here to show that exceptions are indeed being well handled?

Not exactly. the deleted case [test("test SparkContext should not be killed ")](https://github.com/dmlc/xgboost/pull/7836/files#diff-83dc1ab309d0aaa9e7ec44586d99b82cee3c34daa70592fccaca834c6acc27afL121) is only for `kill_spark_context_on_worker_failure` parameter, and there already has the rabbit mock tests https://github.com/dmlc/xgboost/pull/7836/files#diff-83dc1ab309d0aaa9e7ec44586d99b82cee3c34daa70592fccaca834c6acc27afR85

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala (hunk 1 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala:386]: Completely not. I just make this tests more like the test description.

**Review by trivialfis — APPROVED** [review]
